core/src/main/scala/kafka/log/remote/RemoteLogManager.scala

```diff
@@ -444,22 +444,36 @@ class RemoteLogManager(fetchLog: TopicPartition => Option[Log],
       val logInputStream = new RemoteLogInputStream(is)
 
       var batch:RecordBatch = null;
-      while ((batch = logInputStream.nextBatch()) != null && batch.lastOffset < offset) {
+      def fetchBatch() : RecordBatch = {
+        batch = logInputStream.nextBatch()
+        batch
+      }
+
+      // always look for the batch which has the desired offset and set the startPosition as that segments start
+      // location.
+      while (fetchBatch() != null && batch.lastOffset < offset) {
+        // sets the startPos as the starting of the next batch's start location.
         startPos += batch.sizeInBytes
       }
-    } finally {
-      is.close()
-    }
 
-    try {
-      is = remoteLogStorageManager.fetchLogSegmentData(rlsMetadata, startPos, startPos + maxBytes)
-      val buffer = ByteBuffer.allocate(fetchMaxBytes)
-      Utils.readFully(is, buffer)
+      val buffer = ByteBuffer.allocate(maxBytes)
+      var remainingBytes = maxBytes
+      if(batch != null) {
+        remainingBytes -= batch.sizeInBytes()
+        batch.writeTo(buffer)
+      }
+
+      if(remainingBytes > 0) {
+        // input stream is read till (startPos-1) while getting the batch of records earlier.
+        // read the input stream until min of (EOF stream or buffer's remaining size).
+        Utils.readFully(is, buffer)
```

Comment:
What happens if `maxBytes` is reached in `buffer` and the stream `is` is in the middle of a record batch?

Reply:
`MemoryRecords` will return `null` as the next record batch if it does not have the enough remaining bytes in bytebuffer. You can look at [ByteBufferLogInputStream.nextBatch()](ByteBufferLogInputStream) which is used in `MemoryRecords`.

Reply:
I mean, specifically on the `Utils.readFully`, when the remaining data is too large to fit in the remaining space of `buffer`?

Reply:
We should send only `maxBytes` of data as per fetch protocol. `buffer` is created with the same and it would read the minimum of `inputstream` or buffer's capcity. You can read java doc of `Utils.readFully` which is mentioned below.

```    
Read data from the input stream to the given byte buffer until there are no bytes remaining in the buffer or the end of the stream has been reached.
```

Reply:
Actually there is an edge case that we didn't handle properly. 

After Kafka 0.10.1 (KIP-74), the at least one message / batch is always returned, even if its size is larger than the max bytes required by the consumer. 

The logic is at https://github.com/harshach/kafka/blob/342613b8f6e82744415aa60fe5b67fdfbfc677f8/core/src/main/scala/kafka/server/ReplicaManager.scala#L1171

The logic is different for old consumers (before 0.10.1)
https://github.com/harshach/kafka/blob/342613b8f6e82744415aa60fe5b67fdfbfc677f8/core/src/main/scala/kafka/server/KafkaApis.scala#L808

So, for the 1st non-empty topic-partition, we have to check the size of the 1st batch before allocating the buffer.

The minOneMessage parameter also has to be passed from ReplicaManager to RemoteLogManager

Reply:
`minOneMessage` is already passed to RLM and it was sending it to RSM earlier but it was not honored. We can address this scenario in this PR.